File: streamer-v2/src/main/java/eu/tsalliance/streamer/channel/ChannelHandler.java
package eu.tsalliance.streamer.channel;

import eu.tsalliance.streamer.database.MySQL;
import eu.tsalliance.streamer.files.FileHandler;
import lombok.Getter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.ResultSet;
import java.util.HashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ChannelHandler {
    private static final Logger logger = LoggerFactory.getLogger(ChannelHandler.class);
    @Getter private static HashMap<String, Channel> channels = new HashMap<>();
    @Getter private static ExecutorService executorService = Executors.newCachedThreadPool();

    /**
     * Load all channels from database and register them in hashmap
     */
    public static void loadChannels() {
        try {
            if (channels.size() > 0) {
                for (String channelId : channels.keySet()) {
                    stopChannel(channelId);
                }
            }

            channels.clear();

            ResultSet rs = MySQL.query("SELECT uuid, mountpoint, enabled FROM " + MySQL.TABLE_CHANNELS + " WHERE enabled = 1;");
            if (rs != null) {
                do {
                    String uuid = rs.getString("uuid");
                    String mountpoint = rs.getString("mountpoint");
                    boolean enabled = rs.getBoolean("enabled");

                    Channel channel = new Channel(uuid, mountpoint, enabled);
                    registerChannel(channel);
                } while (rs.next());

                logger.info("loadChannels(): Loaded '" + channels.size() + "' channels");
            } else {
                logger.warn("loadChannels(): Could not find channels...");
            }
        } catch (Exception exception) {
            exception.printStackTrace();
        }
    }

    /**
     * Unload a channel
     * @param uuid Channel's uuid
     */
    public static void unloadChannel(String uuid) {
        if(!channelExists(uuid)) {
            return;
        }

        channels.remove(uuid);
        System.gc();
    }

    /**
     * Register a new channel
     * @param channel Channel to register
     */
    public static void registerChannel(Channel channel) {
        if(channelExists(channel.getUuid())) {
            return;
        }

        channels.put(channel.getUuid(), channel);

        if(channel.isEnabled()) {
            startChannel(channel.getUuid());
        }
    }

    /**
     * Get a channel by its uuid
     * @param uuid Channel's uuid
     * @return Channel or null
     */
    public static Channel getChannel(String uuid) {
        return channels.getOrDefault(uuid, null);
    }

    /**
     * Check if a channel is loaded
     * @param uuid Channel's uuid
     * @return True or False
     */
    public static boolean channelExists(String uuid) {
        return channels.containsKey(uuid);
    }

    /**
     * Get the state of a channel
     * @param uuid Channel's uuid
     * @return ChannelState or null
     */
    public static ChannelState getState(String uuid) {
        if(!channelExists(uuid)) {
            return null;
        }

        return getChannel(uuid).getChannelState();
    }

    /**
     * Check if the Channel's state is not STATE_OFFLINE
     * @param uuid Channel's uuid
     * @return True or False
     */
    public static boolean isRunning(String uuid) {
        return getState(uuid) != ChannelState.STATE_OFFLINE;
    }

    /**
     * Start a channel
     * @param uuid Channel's uuid
     */
    public static void startChannel(String uuid) {
        if(!channelExists(uuid)) {
            logger.warn("startChannels(): Could not start channel by id '" + uuid + "': Channel not found");
            return;
        }

        if(isRunning(uuid)) {
            logger.warn("startChannels(): Could not start channel by id '" + uuid + "': Channel already running");
            return;
        }

        executorService.execute(getChannel(uuid));
    }

    /**
     * Stop a channel
     * @param uuid Channel's uuid
     */
    public static void stopChannel(String uuid) {
        if(!channelExists(uuid)) {
            logger.warn("restartChannel(): Could not restart channel '" + uuid + "': Channel not found");
            return;
        }

        if(!isRunning(uuid)) {
            startChannel(uuid);
            return;
        }

        Channel channel = channels.get(uuid);

        try {
            channel.shutdown();
            while(isRunning(uuid)) {
                Thread.sleep(1000);
            }
            Thread.sleep(500);
        } catch (Exception ex) {
            logger.error("restartChannel(): An error occured when stopping a channel: " + ex.getMessage());
        }
    }

    /**
     * Restart a channel
     * @param uuid Channel's uuid
     */
    public static void restartChannel(String uuid) {
        if(!channelExists(uuid)) {
            logger.warn("restartChannel(): Could not restart channel '" + uuid + "': Channel not found");
            return;
        }

        if(!isRunning(uuid)) {
            startChannel(uuid);
            return;
        }

        try {
            stopChannel(uuid);
            startChannel(uuid);
        } catch (Exception ex) {
            ex.printStackTrace();
            logger.error("restartChannel(): An error occured when restarting a channel: " + ex.getMessage());
        }
    }

    /**
     * Delete a channel locally
     * @param uuid Channel to delete
     */
    public static void deleteChannel(String uuid) {
        if(!channelExists(uuid)) {
            return;
        }

        Channel channel = getChannel(uuid);

        // Shutdown channel if running
        if(isRunning(uuid)) {
            stopChannel(uuid);
        }

        unloadChannel(uuid);
        FileHandler.deleteChannelDirectory(channel);
    }

    /**
     * Update a specific channel
     * @param uuid Uuid of target
     * @param updatedChannel Updated data
     */
    public static void updateChannel(String uuid, Channel updatedChannel) {
        if(!channelExists(uuid)) {
            return;
        }

        Channel oldChannel = getChannel(uuid);

        // Only update locally, if mountpoint has changed
        if(!oldChannel.getMountpoint().equals(updatedChannel.getMountpoint())) {
            if(isRunning(uuid)) {
                stopChannel(uuid);
            }

            unloadChannel(uuid);

            FileHandler.renameChannelDirectory(oldChannel, updatedChannel);
            logger.info("updateChannel(): Updated channel '" + oldChannel.getMountpoint() + "' >> "+updatedChannel.getMountpoint() + " - Enabled: "+updatedChannel.isEnabled());

            if(updatedChannel.isEnabled()) {
                channels.put(uuid, updatedChannel);
                startChannel(uuid);
            }
        }

    }

    /**
     * Send channel state updates to socket server
     */
    public static void notifyAllChannelStates() {
        for(Channel channel : channels.values()) {
            channel.notifyChannelStateChange();
        }
    }

    /**
     * Send channel state updates to socket server
     */
    public static void notifyAllChannelInfos() {
        for(Channel channel : channels.values()) {
            channel.notifyChannelInfoChange();
        }
    }

    /**
     * Send channel state updates to socket server
     */
    public static void notifyAllChannelHistories() {
        for(Channel channel : channels.values()) {
            channel.notifyChannelHistoryChange();
        }
    }
}

File: nodeserver/src/main/kotlin/live/tsradio/nodeserver/handler/ChannelHandler.kt
package live.tsradio.nodeserver.handler

import live.tsradio.nodeserver.channel.Channel
import live.tsradio.nodeserver.files.Filesystem
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

object ChannelHandler {
    private val logger: Logger = LoggerFactory.getLogger(ChannelHandler::class.java)

    private val channels: HashMap<UUID, Channel> = HashMap()
    val restartTries: HashMap<UUID, Int> = HashMap()

    private val executorService: ExecutorService = Executors.newFixedThreadPool(Filesystem.preferences.channels.max)

    fun set(channel: Channel) {
        if(this.channels.containsKey(channel.data.id)) {
            this.channels[channel.data.id]!!.data = channel.data
        } else {
            this.channels[channel.data.id] = channel
        }
    }

    fun startChannel(uuid: UUID) {
        if(!channels.containsKey(uuid)) {
            logger.warn("Cannot start non-existent channel")
            return
        }

        if(getRunningChannels().size >= Filesystem.preferences.channels.max) {
            logger.warn("The maximum of simultaneously running channels reached!")
            return
        }

        val nodeChannel = channels[uuid]!!
        if(channelIsRunning(uuid)) {
            logger.warn("Channel '${nodeChannel.name}' is already running!")
            return
        }

        nodeChannel.execute(executorService)
    }

    fun channelExists(uuid: UUID): Boolean {
        return this.channels.containsKey(uuid)
    }

    fun channelIsRunning(uuid: UUID): Boolean {
        return if(this.channels.containsKey(uuid)) {
            this.channels[uuid]!!.isRunning
        } else {
            false
        }
    }

    fun getChannel(uuid: UUID): Channel? {
        return this.channels[uuid]
    }

    fun getRunningChannels(): ArrayList<Channel> {
        return this.channels.values.filter { it.isRunning }.toCollection(ArrayList())
    }
    fun getConfiguredChannels(): ArrayList<Channel> {
        return this.channels.values.filter { !it.isRunning }.toCollection(ArrayList())
    }
    /*fun startChannel(channelName: String) {
        val channel = getChannelByNameLocally(channelName) ?: return
        startChannel(configuredChannels[channel.data.id])
    }

    //TODO: Test
    private fun stopChannel(channel: Channel?, forceStop: Boolean) {
        if(channel == null) {
            logger.warn("Cannot stop non-existent channel")
            return
        }

        channel.cancel(forceStop)
        channel.join()

        // TODO: Trigger Channel Status Change
        //channel.channelInfo.update()
    }
    fun stopChannel(channelName: String, forceStop: Boolean) {
        val channel = getChannelByNameLocally(channelName)?: return
        stopChannel(activeChannels[channel.data.id],forceStop)
    }
    fun restartChannel(channelName: String) {
        val channel = getChannelByNameLocally(channelName)?: return

        stopChannel(channel, true)
        while (activeChannels.containsKey(channel.data.id)){
            // Wait
            Thread.sleep(1000)
        }
        startChannel(channel)
    }

    fun createChannel(channel: Channel){
        /*try {
            MySQL.insert(MySQL.tableChannels, channel.toContentValues())
            logger.info("Channel '${channel.data.name}' created.")
            notifyChannelUpdated(channel)
        } catch (ex: Exception) {
            logger.info("Creating channel '${channel.data.name}' failed: ${ex.message}")
        }*/
    }

    fun deleteChannel(channel: Channel) {
        /*try {
            if(isChannelActiveByID(channel.data.id)) {
                logger.info("Preparing channel '${channel.data.name}' for deletion.")
                channel.cancel(true)
                channel.join()
                activeChannels.remove(channel.data.id)
            }
            configuredChannels.remove(channel.data.id)
            MySQL.delete(MySQL.tableChannels, "nodeID = '${Filesystem.preferences.node.clientID}' AND id = '${channel.data.id}'")
            notifyChannelRemoved(channel.data.name)
            logger.info("Channel '${channel.data.name}' deleted.")
        } catch (ex: Exception) {
            logger.info("Deleting channel '${channel.data.name}' failed: ${ex.message}")
        }*/
    }

    fun editChannel(channelName: String, channelUUID: String, channel: Channel) {
        /*try {
            MySQL.update(MySQL.tableChannels, "nodeID = '${Filesystem.preferences.node.clientID}' AND id = '$channelUUID'", channel.toContentValues())
            notifyChannelUpdated(channel)
            logger.info("Channel '$channelName' edited.")
        } catch (ex: Exception) {
            logger.info("Editing channel '$channelName' failed: ${ex.message}")
        }*/
    }

    fun channelExistsByName(channelName: String): Boolean {
        /*configuredChannels.values.forEach { if(it.data.name == channelName) return true }
        return try {
            MySQL.exists(MySQL.tableChannels, "name = '$channelName'")
        } catch (ignored: Exception) {
            ignored.printStackTrace()
            false
        }*/
        return false
    }
    fun channelExistsByMount(mountpoint: String): Boolean {
        /*configuredChannels.values.forEach { if(it.data.mountpoint == mountpoint) return true }
        return try {
            MySQL.exists(MySQL.tableChannels, "mountpoint = '$mountpoint'")
        } catch (ignored: Exception) {
            false
        }*/
        return false
    }
    fun mysqlResultToChannel(result: ResultSet): Channel? {
        /*val genres = ArrayList<String>()
        val jsonArray = JsonParser().parse(result.getString("genres")).asJsonArray

        jsonArray?.forEach {
            genres.add(it.asString)
        }

        val dataPacket = ChannelDataPacket(
                result.getString("id"),
                result.getString("nodeID"),
                result.getString("name"),
                result.getString("description"),
                result.getString("creatorID"),
                result.getString("mountpoint"),
                result.getString("playlistID"),
                genres,
                result.getBoolean("featured"),
                result.getBoolean("listed"),
                result.getBoolean("playlistShuffle"),
                result.getBoolean("playlistLoop"),
                result.getInt("priority"),
                ChannelInfoPacket(result.getString("id"))
        )

        return Channel(dataPacket)*/
        return null
    }
    fun getChannelOnNodeByName(channelName: String): Channel? {
        /*val result = MySQL.get(MySQL.tableChannels, "nodeID = '${Filesystem.preferences.node.clientID}' AND name = '$channelName'", ArrayList(listOf("*")))

        return if(result != null && result.next()) {
            mysqlResultToChannel(result)
        } else {
            null
        }*/
        return null
    }
    private fun getChannelByNameLocally(channelName: String): Channel? {
        var channel: Channel? = null
        for(chan in configuredChannels.values){
            if(chan.data.name == channelName) {
                channel = chan
                break
            }
        }
        return channel
    }
    fun isChannelActiveByID(channelID: String): Boolean {
        return activeChannels.containsKey(channelID)
    }


    override fun onChannelReady(channel: Channel) {
        logger.info("Channel '${channel.data.name}' is ready")
    }

    override fun onChannelDone(channel: Channel) {
        if(restartTries.containsKey(channel.data.id)) {
            restartTries.remove(channel.data.id)
        }
    }

    fun resetRestartTries(channel: Channel){
        if(restartTries.containsKey(channel.data.id)) {
            restartTries.remove(channel.data.id)
        }
    }

    override fun onChannelStop(channel: Channel, reason: Int) {
        activeChannels.remove(channel.data.id)
        Thread.sleep(10)

        if(reason == REASON_CHANNEL_EXCEPTION && Filesystem.preferences.channels.autorestart) {
            // Restart channel
            var triesUsed = restartTries.getOrDefault(channel.data.id, 0)
            val delay: Long = (Filesystem.preferences.channels.restartDelay*1000).toLong()
            triesUsed += 1

            when {
                triesUsed == 1 -> {
                    logger.warn("Channel '${channel.data.name}' was shut down. Restarting it in ${delay/1000}sec [$triesUsed/${Filesystem.preferences.channels.restartTries}]")

                    Thread.sleep(delay)
                    restartTries[channel.data.id] = triesUsed
                    startChannel(channel.data.name)
                }
                triesUsed > Filesystem.preferences.channels.restartTries -> {
                    logger.error("Could not restart channel '${channel.data.name}' after ${triesUsed-1} tries.")
                    restartTries.remove(channel.data.id)
                }
                else -> {
                    // Execute with delay
                    logger.warn("Trying to restart channel '${channel.data.id}' in ${delay/1000}sec [$triesUsed/${Filesystem.preferences.channels.restartTries}]")

                    Thread.sleep(delay)
                    restartTries[channel.data.id] = triesUsed
                    startChannel(channel.data.name)
                }
            }
        } else {
            logger.info("Channel '${channel.data.name}' shut down.")
        }
    }

    override fun newThread(r: Runnable): Thread {
        return Thread(r, "channel-thread-${activeChannels.size+1}")
    }



    fun notifyChannelUpdated(channel: Channel){
        if(!configuredChannels.containsKey(channel.data.id)) {
            configuredChannels[channel.data.id] = channel

            if(Filesystem.preferences.channels.autostart) {
                startChannel(channel)
            }
            return
        }

        configuredChannels[channel.data.id]!!.liveUpdate(channel)
        if(activeChannels.containsKey(channel.data.id)) {
            activeChannels[channel.data.id]!!.liveUpdate(channel)
        }
    }

    fun notifyChannelRemoved(channelID: String){
        if(configuredChannels.containsKey(channelID)) configuredChannels.remove(channelID)
        if(activeChannels.containsKey(channelID)) {
            stopChannel(channelID, true)
        }
    }*/
}